super-ddf.c

```diff
@@ -2197,6 +2197,30 @@ static void uuid_from_super_ddf(struct supertype *st, int uuid[4])
 		uuid_from_ddf_guid(ddf->anchor.guid, uuid);
 }
 
+int
+getinfo_phys_disk_state(struct phys_disk_entry *pde)
+{
+	int state = 0;
+	if (!pde) {
+		state |= (1 << MD_DISK_FAULTY);
+		return state;
+	}
+
+	// if disk is MISSING or FAILED it is considered FAULTY.
+	// Otherwise if ONLINE, mark as ACTIVE.
+	int stt = be16_to_cpu(pde->state);
+	if ((stt & (DDF_Failed|DDF_Missing|DDF_Online)) == DDF_Online)
+		state |= (1 << MD_DISK_ACTIVE);
+	else
+		state |= (1 << MD_DISK_FAULTY);
+
+	// if disk is ONLINE and is NOT REBUILDING, then mark it IN_SYNC.
+	if ((stt & (DDF_Failed|DDF_Missing|DDF_Rebuilding|DDF_Online)) == DDF_Online)
+		state |= (1 << MD_DISK_SYNC);
+
```

Comment:
This logic seems slightly different than the existing code, which I assume is ok. It seems alright I guess. The old code would have considered a disk that's !missing and !failed as `ACTIVE` *and* `SYNC`, which was part of the issue, I assume?

Reply:
This is actually the main actual change of behavior, and is what provides the fix here. By *NOT* having it `SYNC` if it's rebuilding, this fixes the problem.